Extend set_value for use on check boxes, radio buttons and drop-down options

graeme-defty opened the pull request.

graeme-defty commented:
Sorry for the clumsy gitting. (I tend to use SVN).  I will try to do better in future.

PragTob commented:
@graeme-defty don't worrty about the git-ing too much, we usually squash commits which means that all of this will end up as a single commit on master :+1:

graeme-defty commented:
:-)  Thanks

On 13 May 2017 at 19:52, Tobias Pfeiffer <notifications@github.com> wrote:

> @graeme-defty <https://github.com/graeme-defty> don't worrty about the
> git-ing too much, we usually squash commits which means that all of this
> will end up as a single commit on master 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keathley/wallaby/pull/237#issuecomment-301246325>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAK80JecrgEx7jdt3-uUhIJIeTeGiiZrks5r5adwgaJpZM4NZ9zI>
> .
>

PragTob commented on .gitignore:
```diff
@@ -3,6 +3,7 @@
 /deps
 erl_crash.dump
 *.ez
+mix.lock
```
We don't wanna ignore mix.lock :)

In the Ruby world it's common and I used to do it, but I argued with José once and he had (of course!) very valid points, e.g. a common experience for maintainers etc. - so we'd wanna keep this around :)

graeme-defty replied:
I just read some of the discussions on this. Thanks for the pointer. (I should have guessed that it was needed by the fact that nobody ignored it yet - duh!)

PragTob commented on integration_test/cases/browser/set_value_test.exs:
```diff
@@ -0,0 +1,71 @@
+
+      assert page
+      |> find(Query.text_field("email"))
+      |> has_value?("c@keathley.io")
```
this test confuses me a bit, it doesn't seem to call `set_value` at all?

graeme-defty replied:
I addressed this in response to your other point. Should I just remove this test or try to make it fit set_value?

PragTob replied:
I think just removing it should be fien

PragTob commented on integration_test/cases/browser/set_value_test.exs:
```diff
@@ -0,0 +1,71 @@
+  describe "set_value/3" do
+    test "accepts a query", %{page: page} do
+      page
+      |> send_keys(Query.text_field("Name"), ["Chris", :tab, "c@keathley.io"])
```
Is there something specific here that we'd wanna use `send_keys` over `fill_in` for filling in the form fields?

graeme-defty replied:
I copied the test code from sends_keys and seem to have left this in as a hangover from that.  To be honest, I would not have a test for  "accepts a query", as I could not see what it was doing over and above the other tests. I kept it to stay consistent, but did not tailor it to set_value. Mea culpa.

PragTob commented on integration_test/cases/browser/set_value_test.exs:
```diff
@@ -0,0 +1,71 @@
+
+      assert page
+      |> find(Query.text_field("Name"))
+      |> has_value?("Chris")
```
sorry we don't have a code linter setup yet, but I think indentation wise we'd like this rather to be:

```
assert page
       |> call
       |> call
```

(goes for all stuff here) :)

PragTob replied:
This also makes me think that maybe we should have a `value` option for queries (if we don't have that already) so that this could just be: `assert_has(parent, Query.text_field("Name", value: "Chris"))` but that's for another time :)

graeme-defty replied:
OK - thanks. Noted for the future.

PragTob commented on integration_test/cases/browser/set_value_test.exs:
```diff
@@ -0,0 +1,71 @@
+      |> set_value(Query.option("Select Option 2"), :selected)
+      |> find(Query.option("Select Option 1"))
+      |> selected?()
+    end
```
Good amount of test coverage imo, thanks! :+1:

graeme-defty commented:
Thanks for the comments. Most importantly, if Nala likes it them I am happy!

What is the procedure now - do I make changes addressing your comments and re-submit the PR?

PragTob commented:
@graeme-defty you make the changes addressing the comments, then you make a new commit with them and just push them to the branch again. We'll get them here and everything should be fine :)

graeme-defty commented:
OK - done!

aaronrenner commented on lib/wallaby/browser.ex:
```diff
@@ -392,12 +392,37 @@ defmodule Wallaby.Browser do
+#    IO.warn "set_value/2 has been deprecated. Please use Element.set_value/2"
+#
+#    Element.set_value(element, value)
+#  end
```
Could the commented out code be removed now?

@keathley, @PragTob  Should this change be on `Element.set_value` instead `Browser.set_value`?

keathley replied:
I think we want to keep this in the `Browser` module since it accepts queries. We will want to remove the commented out code.

keathley commented on lib/wallaby/browser.ex:
```diff
@@ -392,12 +392,37 @@ defmodule Wallaby.Browser do
   @doc """
   Sets the value of an element.
```
Since we're explicitly looking for `:selected` and `:unselected` we'll want to call that out in the documentation.

graeme-defty commented:
Ouch!  Yes, I should have done this.

On 15 May 2017 at 20:10, Aaron Renner <notifications@github.com> wrote:

> *@aaronrenner* commented on this pull request.
> ------------------------------
>
> In lib/wallaby/browser.ex
> <https://github.com/keathley/wallaby/pull/237#discussion_r116484461>:
>
> > @@ -392,12 +392,37 @@ defmodule Wallaby.Browser do
>    @doc """
>    Sets the value of an element.
>    """
> -  @spec set_value(element, any()) :: element
> +  @spec set_value(parent, Query.t, Element.value) :: parent
> +#  @spec set_value(element, any()) :: element
> +
> +#  def set_value(element, value) do
> +#    IO.warn "set_value/2 has been deprecated. Please use Element.set_value/2"
> +#
> +#    Element.set_value(element, value)
> +#  end
>
> Could the commented out code be removed now?
>
> @keathley <https://github.com/keathley>, @PragTob
> <https://github.com/pragtob> Should this change be on Element.set_value
> instead Browser.set_value?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keathley/wallaby/pull/237#pullrequestreview-38108830>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAK80KzkguzFYaR83I1l-lw-i0FTWaxAks5r6E7ZgaJpZM4NZ9zI>
> .
>

graeme-defty commented:
Documentation updated as requested

PragTob commented:
I think this clashed with a merge to master now where `selected?` was removed from Browser #240 - so best to merge that inrebase against it and fix the tests relying on it :)

Sorry for the trouble @graeme-defty if you can't get to it one of us probably will :)

graeme-defty commented:
Let me try - my gitting skills could use the workout  :-s

graeme-defty commented:
Done!  I still officially loathe git though. How on earth has it won over BZR? (Though I have to admit --rebase is pretty cool. Once I eventually got the incantation right, it just ... well ... worked!  Maybe I will get used to it.  ;-)

keathley commented:
@graeme-defty No worries on the git stuff. I cleaned up your branch and merged everything here: https://github.com/keathley/wallaby/commit/e9f4e1a071df3d9b75dcd86c46f1e18bf810ea42
